python/pyspark/sql/functions.py

```diff
@@ -2840,6 +2840,360 @@ def from_csv(col, schema, options={}):
     return Column(jc)
 
 
+def _invoke_higher_order_function(name):
+    """
+    Given name of the expression corresponding to a higher order SQL function,
+    return a function that takes
+
+    - a list of columns
+    - a list of tuples ((*Column) -> Column, Iterable[int])
+      where the second element represent allowed arities
+
+    invokes expression, and wraps result in a Column
+    (first Scala one, then Python).
+    """
+
+    def _(cols, funs):
```

Comment:
@zero323, I know you were trying to logically separate the actual function but how about we just avoid having a nested function here? I think it also makes sense not to have the nested function here and I personally think it's a bit confusing in particular to people not quite used to Python.

Reply:
@HyukjinKwon 

> I know you were trying to logically separate the actual function but how about we just avoid having a nested function here?

Not a problem. It felt natural to think about the data flow this way, but I am not very attached to this approach.